README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# dbox

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![CRAN
status](https://www.r-pkg.org/badges/version/dbox)](https://cran.r-project.org/package=dbox)
[![Travis build
status](https://travis-ci.org/hwkmth/dbox.svg?branch=master)](https://travis-ci.org/hwkmth/dbox)
<!-- badges: end -->

`dbox` is a package for visualizing continuous distributions with a
combined kernel density plot and box plot. This allows for a
simultaneous view of overall distribution shape in the density plot as
well as important quantiles and extreme values in the box plot. Box plot
outliers are jittered to improve outliers density visualization. An
optional density plot of an idealized normal distribution with the same
parameters is also available for comparison.

## Installation

You can install dbox with: `devtools::install_github('hwkmth/dbox')` and
use `help(dbox)` for more information.

## Examples

``` r
library(dbox)

ndensity <- rnorm(5000)
dbox(ndensity)
```

<img src="man/figures/README-examples-1.png" width="100%" />

``` r

rskew <- 1.5^rnorm(5000) * 5
dbox(rskew, color = "orange2", fill = FALSE,
     normal = TRUE, color_norm = "steelblue2")
```

<img src="man/figures/README-examples-2.png" width="100%" />

``` r

diamonds <- ggplot2::diamonds
diamonds$lprice <- log(diamonds$price)
dbox(diamonds$lprice[diamonds$cut == "Ideal"],
     label = "Log Price: Ideal Cut Diamonds",
     color = "orange2",
     normal = TRUE,
     color_norm = "steelblue2")
```

<img src="man/figures/README-examples-3.png" width="100%" />


NEWS
dbox v0.0.4 (Release date: 2019-07-05)
=============

Changes:
* Weights, in general, actually work.
* Relatedly: now depends on Hmisc.
* "w" argument renamed "weights" for clarity.
* When weights are supplied, stop if any complete case of x has a missing
  weight, and vice versa.
* When weights are supplied, comparison normal distribution is constructed
  using weights.
* Added <plot> option, defaulting to TRUE. If FLASE, plot output can be
  assigned to an object.
* Added "guilford" dataset to demonstrate (and test) use of weights.
* Changed conditional plotting of normal comparison, fill, and outliers to
  if/else statements ins
* Changed default line weight to 1 (from 1.1).


dbox v0.0.3 (Release date: 2019-07-02)
=============

Changes:
* Omit missings at beginning of function; use completed cases only for all
  other processes.
* <normal> option plots an *idealized* normal distribution to +/- 3.1 std dev,
  rather than a simulated random normal distribution. Prevents non-normal
  distribution when n is small.
* Added option to specify variable name via <label> argument.
* Better parsing of variable name via sys.call() when <label> is not specififed.
* Include linetype in legend.
* <fill> option affects boxplot fill as well as density plot fill.


dbox v0.0.2 (Release date: 2019-06-29)
=============

Changes:
* Renamed package and main function to dbox.
* Switched to ggplot native stat_boxplot(), rather than combining several
  geom_segment() & annotate() into semi-manual boxplot. Yields a significant
  speed improvement.
* Switched to geom_path() rather than geom_stat() for density plots to get
  intial orientation the same as stat_boxplot().
* Now computes densities with n=512, rather than n=length(x). Yields a
  significant speed improvement.


R/test_dists.R
#' Generate data to test distribution visualizations.
#'
#' Generates a dataframe of several random distributions testing.
#' Distributions include: normal, right- and left-skewed, uniform, bimodal,
#' absolute value sine, and a "weird" distribution of concatenated
#' chi-squared's with random parameters.
#'
#' @param n Number of observations.
#' @import stats
#' @export
#' @examples
#' testing <- test_dists(5000)


test_dists <- function(n) {
  normal <- rnorm(n) / 1.1
  rskew <- 1.5^(rnorm(n)) * 5
  lskew <- log(rnorm(n)^2)
  bimodal <- c(sample(lskew, n/2), sample(rskew, n/2))
  uniform <- runif(n, -3, 3)
  sine <- abs(sin(seq(-pi, pi, 2*pi/n)))
  sine <- sine[1:n]
  weird <- NULL
  while (length(weird) < n) {
    weird <- append(
      weird,
      rchisq(
        sample(1:(100), 1),
        sample(1:5000, 1),
        sample(1:500, 1)
      ) ^ (1/(sample(3:9, 1)))
    )
  }
  weird <- weird[1:n]

  dists <- data.frame(
    normal,
    rskew,
    lskew,
    uniform,
    bimodal,
    sine,
    weird
  )

  return(dists)
}


R/data.R
#' Test Data for Distribution Visualizations
#'
#' A set of random distributions for visualization tests, genereated by
#' densitybox::test_dists(). For for details see documentation there.
#'
#' @format A data frame of 10,000 observations and 7 variables.
#' \describe{
#'   \item{normal}{Random normal}
#'   \item{rskew}{Right-skewed through exponentiation}
#'   \item{lskew}{Left-skewed through log transformation & exponentiation}
#'   \item{uniform}{Uniform on [-3,3]}
#'   \item{bimodal}{Bimodal sampled from rskew & lskew}
#'   \item{sine}{Sampled from absolute value sine function}
#'   \item{weird}{Concatenated Chi-squared distributions with random parameters}
#' }
'dbox_dists'

#' Guilford County, NC 2017 Income Data
#'
#' 3,352 observations plus survey weights of 2017 Guilford County, NC personal
#' income. Source: American Community Survey, via IPUMS USA.
#'
#' @format A data frame of 3,352 observations and 2 variables.
#' \describe{
#   \item{income}{Personal income in USD for persons age 16+ with income.}
#   \item{perwt}{Person weight: unscaled survey weight for a 1% sample.}
#' }
'guilford'


R/densitybox.R
#' Combined density/box plots.
#'
#' @param x A continuous numeric vector.
#' @param color Fill and line color of both density and box plots.
#' @param alpha Fill transparency of both density and box plots.
#' @param linewt Weight of density plot outline.
#' @param linewt_box Weight of boxplot lines.
#' @param normal Logical. Include a simulated normal/Gaussian density line plot with the same parameters as x? Defaults to FALSE.
#' @param color_norm Line color of the optional simulated normal plot.
#' @param linewt_norm Wieght of the optional simulated normal plot.
#' @import ggplot2
#' @importFrom stats density rnorm sd
#' @importFrom graphics boxplot
#' @export
#' @examples
#' ndensity <- rnorm(5000)
#' dbox(ndensity)

dbox <- function(x,
                 color = "black",
                 alpha = 0.2,
                 linewt = 1.1,
                 linewt_box = 0.8,
                 normal = FALSE,
                 color_norm = "black",
                 linewt_norm = 1.1
                 ) {

  bpstats <- as.vector(boxplot(x, plot = FALSE)$stats)
  outliers <- boxplot(x, plot = FALSE)$out
  ymin <- -(max(density(x)$y)/20)
  ymax <- -(max(density(x)$y)/4)
  varname <- deparse(substitute(x))

  p <- ggplot() +
    # density fill
    geom_density(
      aes(x = x),
      fill = color,
      color = NA,
      alpha = 0.2
    ) +
    # density upper border
    stat_density(
      geom = "line",
      aes(x = x, color = varname),
      size = linewt
    ) +
    # boxplot central box
    annotate(
      "rect",
      xmin = bpstats[2],
      xmax = bpstats[4],
      ymin = ymin,
      ymax = ymax,
      fill = color,
      alpha = 0.2,
      color = color,
      size = linewt_box
    ) +
    # boxplot median line
    geom_segment(
      data = data.frame(x = x),
      x = bpstats[3],
      xend = bpstats[3],
      y = ymin,
      yend = ymax,
      color = color,
      size = linewt_box
    ) +
    # boxplot left whisker
    geom_segment(
      data = data.frame(x = x),
      x = bpstats[1],
      xend = bpstats[2],
      y = (ymax + ymin) / 2,
      yend = (ymax + ymin) / 2,
      color = color,
      size = linewt_box
    ) +
    # boxplot right whisker
    geom_segment(
      data = data.frame(x = x),
      x = bpstats[4],
      xend = bpstats[5],
      y = (ymax + ymin) / 2,
      yend = (ymax + ymin) / 2,
      color = color,
      size = linewt_box
    ) +
    # theme
    theme_minimal() +
    theme(
      axis.title = element_blank(),
      axis.text.y = element_blank(),
      panel.grid.major.y = element_blank(),
      panel.grid.minor.y = element_blank(),
      panel.grid.minor.x = element_blank(),
      legend.position = "bottom"
    )

  # outliers, if any
  o <- geom_jitter(
    data = data.frame(x = outliers),
    aes(
      x = outliers,
      y = (ymax + ymin) / 2
    ),
    color = color,
    alpha = 0.2,
    height = abs((ymax + ymin) / 5)
  )

  # normal density comparison, optional
  n <- stat_density(
    geom = "line",
    aes(x = rnorm(length(x), mean(x), sd(x)),
      color = "simulated normal"
    ),
    linetype = "twodash",
    size = linewt_norm
  )

  # conditional plotting
  if (normal == TRUE & length(outliers != 0)) {
    p + n + o + scale_color_manual(name = "", values = c(color, color_norm))
  } else if (normal == FALSE & length(outliers != 0)) {
    p + o + scale_color_manual(name = "", values = c(color))
  } else if (normal == TRUE & length(outliers == 0)) {
    p + n + scale_color_manual(name = "", values = c(color, color_norm))
  } else
    p + scale_color_manual(name = "", values = c(color))

}


R/dbox.R
#' Combined density/box plots.
#'
#' @param x A continuous numeric vector.
#' @param weights A vector of weights. Must be either NULL (the default) or the
#'   same length as `x`. All complete cases of `x` must have a non-missing
#'   value for `weights`.
#' @param plot Logical, whether to plot immediately. Default is TRUE. If FALSE,
#'   returns the plot, which can be assigned to an object.
#' @param fence_coef Defines the length of boxplot whiskers & defines
#'   outliers. Defaults to the typical value of 1.5.
#' @param normal Logical. Include a simulated normal/Gaussian density
#'   plot? Gets parameters from `x` Defaults to FALSE.
#' @param label A character string supplying a variable name to the legend.
#'   The default, NULL, passes the argument supplied to `x` to the legend.
#' @param color Fill and line color of both density and box plots.
#' @param alpha Fill transparency of both density and box plots.
#' @param lwt Weight of density plot outline.
#' @param ltype Line type (pattern) of density plot outline. Default is solid.
#'   For other options, see ?linetype.
#' @param fill Logical. Fill the density plot? Gets color from `color` and
#'   alpha from `alpha`. Defaults to TRUE
#' @param color_norm Line color of the optional simulated normal plot.
#' @param lwt_norm Weight of the optional simulated normal plot.
#' @param ltype_norm Line type simulated normal. Default is twodash.
#' @import ggplot2
#' @importFrom Hmisc wtd.mean wtd.var wtd.quantile
#' @importFrom stats density rnorm sd quantile IQR
#' @importFrom graphics boxplot
#' @export
#' @examples
#' # basic
#' ndensity <- rnorm(5000)
#' dbox(ndensity)
#'
#' # Using colors & comparison normal density plot
#' rskew <- 1.5^rnorm(5000) * 5
#' dbox(rskew, color = "orange2", fill = FALSE,
#'      normal = TRUE, color_norm = "steelblue2")
#'
#' # Assignment to an object using `plot = FALSE` and using a variable label
#' setosa <- subset(iris, Species == "setosa")
#' setosa_petLn <- dbox(setosa$Petal.Length, plot = FALSE,
#'                      label = "Setosa Petal Width", color = "orange2",
#'                      normal = TRUE, color_norm = "steelblue2")
#' setosa_petLn
#'
#' # Using weights, a wider boxplot fence coefficient, and a thinner line
#' dbox::guilford
#' dbox(guilford$income)
#' dbox(guilford$income,
#'      weights = guilford$perwt,
#'      fence_coef = 2,
#'      lwt = 0.8)


dbox <- function(x,
                 weights = NULL,
                 plot = TRUE,
                 fence_coef = 1.5,
                 label = NULL,
                 color = "black",
                 alpha = 0.2,
                 lwt = 1,
                 ltype = 1,
                 fill = TRUE,
                 normal = FALSE,
                 color_norm = "black",
                 lwt_norm = 1,
                 ltype_norm = 6
                 ) {

  # variable name as character string
  varname <- ifelse(is.null(label),
                    deparse(as.list(sys.call())[[2]]),
                    label)

  # Input errors/warnings
  if (length(x) == 0) stop("x has no observations")
  if (length(na.omit(x)) == 0) stop("x has no complete cases")
  if (!is.numeric(x)) stop("x must be numeric")
  if (is.factor(x)) {
    warning("x should be a continuous variable. weird results may ensue.")
  }
  if (!is.null(weights)) {
    if (!is.numeric(weights)) stop("weights must be numeric or NULL")
    if(length(weights) != length(x)) stop("length of weights must equal length of x")
    if (isFALSE(all(complete.cases(x) == complete.cases(weights)))) {
      stop("complete cases of x do not match complete cases of weights: check missing values")
    }
  }
  if (!is.logical(fill)) stop("please supply TRUE/FALSE to fill")
  if (!is.logical(normal)) stop("please supply TRUE/FALSE to normal")
  if (!is.null(label)) {
    if (!is.character(label)) stop("label must be a character vector (or NULL)")
    if (length(label) > 1) stop("please supply only one label")
  }


  # complete cases only. cases will match, as checked above.
  x <- na.omit(x)
  if(!is.null(weights)) wts <- na.omit(weights)

  # compute density, conditional on whether weights are supplied
  if (is.null(weights)){
    dens <- density(x)
  } else {
    wts_scaled <- wts / sum(wts)
    dens <- density(x, weights = wts_scaled)
  }

  # create density data frame
  ddf <- data.frame(
    x = dens$x,
    density = dens$y
  )


  # determine outliers
  if (is.null(weights)) {
    p25 <- quantile(x, .25)
    p75 <- quantile(x, .75)
  } else {
    p25 <- wtd.quantile(x, weights = wts, probs = 0.25)
    p75 <- wtd.quantile(x, weights = wts, probs = 0.75)
  }

  iqr <- p75-p25

  outs <- x[x < (p25 - fence_coef * iqr) |
            x > (p75 + fence_coef * iqr)]


  # compute comparison simulated normal distribution & data frame
  # to +/- 3.1 standard deviations
  if (is.null(weights)) {
    meanx <- mean(x)
    sdx <- sd(x)
  } else {
    meanx <- wtd.mean(x, weights = wts)
    sdx <- sqrt(wtd.var(x, weights = wts))
  }

  norm_x <- seq(-3.1, 3.1, length.out = 512) * sdx + meanx
  norm_y <- dnorm(norm_x, meanx, sdx)
  ndf <- data.frame(
    x = norm_x,
    density = norm_y
  )



  ### INDIVIDUAL PLOT ELEMENTS ###

  # desnity line plot
  d <- geom_path(
    data = ddf,
    aes(x = density, y = x, color = varname, linetype = varname),
    size = lwt
  )

  # density fill/area plot
  if (fill) {
    df <- geom_polygon(
      data = ddf,
      aes(x = density, y = x),
      fill = color,
      color = NA,
      alpha = alpha
    )
  } else {
    df <- NULL
  }


  # boxplot
  if (is.null(weights)) {
    b <- stat_boxplot(
      data = NULL,
      aes(y = x),
      coef = fence_coef,
      fill = NA,
      color = color,
      outlier.shape = NA,
      width = .2 * max(ddf$density),
      position = position_nudge(x = -0.15 * max(ddf$density))
    )
  } else {
    b <- stat_boxplot(
      data = NULL,
      aes(y = x, weight = wts),
      coef = fence_coef,
      fill = NA,
      color = color,
      outlier.shape = NA,
      width = .2 * max(ddf$density),
      position = position_nudge(x = -0.15 * max(ddf$density))
    )
  }


  # boxplot fill
  bf <- annotate(
    geom = "rect",
    xmin = -.05 * max(ddf$density),
    xmax = -.25 * max(ddf$density),
    ymin = p25,
    ymax = p75,
    color = NA,
    fill = color,
    alpha = alpha,
  )

  # outliers
  if (length(outs) > 0) {
    o <-  geom_jitter(
      aes(y = outs, x = -0.15 * max(ddf$density)),
      width = .075 * max(ddf$density),
      height = 0,
      color = color,
      alpha = alpha
    )
  } else {
    o <- NULL
  }


  # optional normal density comparison plot
  if (normal) {
    n <- geom_path(
      data = ndf,
      aes(x = density, y = x,
          color = "Comparison Normal Distribution",
          linetype = "Comparison Normal Distribution"),
      size = lwt_norm
    )
  } else {
    n <- NULL
  }


  # key-value pairs as named vectors for scale constructions
  keys <- c(varname, "Comparison Normal Distribution")
  colorvals <- c(color, color_norm)
  names(colorvals) <- keys
  ltypevals <- c(ltype, ltype_norm)
  names(ltypevals) <- keys

  # scales for legend, conditional on `normal`
  if (normal) {
    sc <- scale_color_manual(name = "", values = colorvals)
    sl <- scale_linetype_manual(name = "", values = ltypevals)
  } else {
    sc <- scale_color_manual(name = "", values = color)
    sl <- scale_linetype_manual(name = "", values = ltype)
  }

  # theme
  t <- theme_minimal() +
    theme(
      axis.title = element_blank(),
      axis.text.y = element_blank(),
      panel.grid.major.y = element_blank(),
      panel.grid.minor.y = element_blank(),
      legend.position = "bottom"
    )


  # plot or return
  dboxplot <- ggplot() + df + d + bf + b + n + o + sc + sl + t + coord_flip()

  if (plot) {
    dboxplot
  } else (
    return(dboxplot)
  )

}
